### ISSUE TITLE
needless_pass_by_value should white-list RangeArgument

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang-nursery/rust-clippy#2357]
There are some argument types for which passing them by value is necessary. 

For example if `Vec.drain(..)` would take `RangeArgument` by reference, one would need to write `Vec.drain(&..)` which defeats the purpose of `RangeArgument`.

**Comment by oli-obk** [comment]
I'm confused. Do you have an example that uses RangeArgument and triggers the lint?

**Comment by gnzlbg** [comment]
The implementation of Vec::drain does. Basically it does not make sense to
take RangeArgument by reference unless one wants the &.. syntax instead of
.. syntax.

**Comment by oli-obk** [comment]
repro:

```rust
#![feature(collections_range)]

use std::collections::range::RangeArgument;
use std::collections::Bound::*;

pub struct V<T>(T);

impl<T> V<T> {
    fn len(&self) -> usize { unimplemented!() }
    pub fn drain<R>(&mut self, range: R) -> (usize, usize)
        where R: RangeArgument<usize>
    {
        let len = self.len();
        let start = match range.start() {
            Included(&n) => n,
            Excluded(&n) => n + 1,
            Unbounded    => 0,
        };
        let end = match range.end() {
            Included(&n) => n + 1,
            Excluded(&n) => n,
            Unbounded    => len,
        };
        assert!(start <= end);
        assert!(end <= len);

        (start, end)
    }
}

fn main() {}
```
